### DISCUSSION TITLE
kubernetes: controlling processes per node

### DISCUSSION THREAD

**Discussion by sheevy** [pmodels/mpich#6209] [Q&A]
Is there a way of controlling a number of processes per node via an env variable? Similar to I_MPI_PERNODE in Intel's MPI version. Or can it only be done using `-ppn` command line parameter?

**Comment by hzhou** [comment]
Currently hydra only support setting ppn via command line. It is not difficult to add such env variable though. What is your use case that command line option is insufficient?

**Reply by sheevy** [reply]
I'm trying to add support for MPICH to [mpi-operator](https://github.com/kubeflow/mpi-operator) which is an extention to Kubernetes to enable it to run MPI jobs. Currently it supports Open MPI and Intel MPI but not MPI. There's a code there which controls number of processes per node which is controlled via a environment variable. It's OMPI_MCA_orte_set_default_slots in case of Open MPI and I_MPI_PERHOST. I'm trying to shoehorn MPICH into that way of doing things.
You can see my PR here: https://github.com/kubeflow/mpi-operator/pull/478

**Reply by hzhou** [reply]
Thanks for the background. That is interesting.

The default behavior (-ppn=1) is to assign ranks in a round-robin fashion. So if you have 4 hosts and 8 processes, each hosts will get 2 processes. The same is true for -ppn=2, but they affect the rank ordering. With -ppn=1 (default), the 8 ranks are on node 0, 1, 2, 3, 0, 1, 2, 3; but with -ppn=2, the 8 ranks are on node 0, 0, 1, 1, 2, 2, 3, 3.

So if you don't care about rank ordering but only cares about processes getting distributed to the hosts evenly, then the default should work.

In addition, when you specify hosts file, you can append per-host ppn. For example,
```
host1:2
host2:4
```
So the first 2 ranks go on to host1 and the next 4 ranks go on to host2,  then round-robin. If you intend for uneven rank distribution, you'll need specify it per-host anyway.

**Reply by sheevy** [reply]
Then I misunderstood the meaning of `ppn`. How do I setup how many processes to lanuch per host? I.e. what option to use to have 2 processes per host, as in your example?

**Reply by alculquicondor** [reply]
That sounds different from OpenMPI https://www.open-mpi.org/doc/v4.0/man1/mpirun.1.php#sect3, unless I'm misinterpreting the documentation. It sounds like I can do `mpirun` without `-n` and it would launch `workers*ppn` ranks.

In any case, if we can do it via the hostfile, that should be fine.

**Reply by hzhou** [reply]
MPICH, or hydra (mpiexec) always checks `-n` first for total number of processes. It is non-standard when `-n` is omitted, and mpich and openmpi use different default. MPICH will check for Job scheduler and resource manager for the number of processes. If that fails, default to 1.

When total number of processes are given and a list of hosts provided, hydra use round-robin to distribute processes. The ppn affects how it is distributed. By default, ppn=1, one at a time. Other wise per ppn at a time.

**Reply by alculquicondor** [reply]
> MPICH will check for Job scheduler and resource manager for the number of processes

How? kubernetes is a resource manager, so how can we give that signal to MPICH?

**Reply by hzhou** [reply]
Is there some standard mechanism kubernetes describe its job environment? For example, with slurm, it sets environment variables such as `SLURM_TASKS_PER_NODE`, `SLURM_JOB_NUM_NODES`, etc. Currently, hydra does not know about the kubernetes environement.

**Reply by alculquicondor** [reply]
We could add whatever is needed. But if mpich already reads certain environment variables, that's better, so users don't need to upgrade mpich.

In terms of number of nodes, why is an environment variable necessary? can't it be deduced from the hostfile?

**Reply by hzhou** [reply]
Hostfile is typically static. Different job launches may desire different number of nodes but it is not convenient to update hostfile each time.

**Reply by alculquicondor** [reply]
In kubernetes, each job runs on a brand new set of workers. A worker is a container in the kubernetes cluster.
As a result, we generate a hostfile everytime.

Is there any environment variable we can use that is not prefixed with `SLURM_`?

**Reply by hzhou** [reply]
> We could add whatever is needed. But if mpich already reads certain environment variables, that's better, so users don't need to upgrade mpich.

MPICH first detect which job scheduler then check the corresponding environment variable. It is bad business to fake a different environment. 

For general environment variable, hydra currently checks `HYDRA_HOST_FILE`. So if you don't mind creating a host file for each job launch, then you can set `HYDRA_HOST_FILE` to let hydra use that host file. You can specify per-host ppn in the host file.

**Reply by hzhou** [reply]
Oh sorry, looks like hydra treats ppn the same way as openmpi if you omit `-n`.

**Reply by hzhou** [reply]
Just to establish the correct answer, if the host file is
```
host1:1
host2:4
host3:2
```
then `mpirun command` (omitting `-n`) will launch 7 processes according to the host file.

**Reply by alculquicondor** [reply]
if we omit `-n`, would hydra use all the hosts in the hostfile?

**Reply by hzhou** [reply]
Yes!

**Reply by alculquicondor** [reply]
oops, race-condition between the comments :)

Thanks for your help @hzhou.

I hope we can collaborate in the future if there are things we can do in mpich to support kubernetes better.
